compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TilingInterfaceImpl.cpp

```diff
@@ -1236,6 +1236,99 @@ SmallVector<Range> UnPackOp::getIterationDomain(OpBuilder &builder) {
+// Im2colOp
+//===----------------------------------------------------------------------===//
+
+SmallVector<Range> Im2colOp::getIterationDomain(OpBuilder &builder) {
+  Location loc = getLoc();
+  Value zero = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = builder.create<arith::ConstantIndexOp>(loc, 1);
+  Value dest = getOutput();
+  SmallVector<Range> loopBounds(getOutputRank());
+  for (int dim = 0; dim < getOutputRank(); ++dim) {
+    loopBounds[dim].offset = zero;
+    loopBounds[dim].size = getDimValue(builder, loc, dest, dim);
+    loopBounds[dim].stride = one;
+  }
+  return loopBounds;
+}
+
+SmallVector<utils::IteratorType> Im2colOp::getLoopIteratorTypes() {
+  SmallVector<utils::IteratorType> iteratorTypes(getOutputRank(),
+                                                 utils::IteratorType::parallel);
+  return iteratorTypes;
+}
+
+FailureOr<TilingResult>
+Im2colOp::getTiledImplementation(OpBuilder &builder,
+                                 ArrayRef<OpFoldResult> offsets,
+                                 ArrayRef<OpFoldResult> sizes) {
+  Location loc = getLoc();
+  auto one = builder.getIndexAttr(1);
+  auto zero = builder.getIndexAttr(0);
+
+  ReifiedRankedShapedTypeDims reifiedInputShapes;
+  SmallVector<OpFoldResult> inputOffsets(getInputRank(), zero);
+  SmallVector<OpFoldResult> inputSizes = getDims(builder, loc, getInput());
+
+  // Set batch offsets and sizes for input
+  for (auto [idx, dim] : llvm::enumerate(getBatchPos())) {
+    inputOffsets[dim] = offsets[idx];
+    inputSizes[dim] = sizes[idx];
+  }
+
+  SmallVector<OpFoldResult> inputStrides(getInputRank(), one);
+  Value inputSlice = getSlice(builder, loc, getInput(), inputOffsets,
+                              inputSizes, inputStrides);
+  SmallVector<OpFoldResult> outputStrides(getOutputRank(), one);
+  Value outputSlice =
+      getSlice(builder, loc, getOutput(), offsets, sizes, outputStrides);
+
+  SmallVector<Type, 4> resultTypes;
+  if (hasPureTensorSemantics()) {
+    resultTypes.push_back(outputSlice.getType());
+  }
+
+  AffineExpr d0, d1;
+  bindDims(getContext(), d0, d1);
+  auto map = AffineMap::get(2, 0, {d0 + d1}, getContext());
+  OpFoldResult kTileOffset = offsets.back();
+  OpFoldResult kOpOffset = getMixedKOffset()[0];
+  OpFoldResult kOffset = affine::makeComposedFoldedAffineApply(
+      builder, loc, map, {kTileOffset, kOpOffset});
+  OpFoldResult mTileOffset = offsets[offsets.size() - 2];
+  OpFoldResult mOpOffset = getMixedMOffset()[0];
+  OpFoldResult mOffset = affine::makeComposedFoldedAffineApply(
+      builder, loc, map, {mTileOffset, mOpOffset});
+
+  SmallVector<Value> operands = {inputSlice, outputSlice};
+  operands.append(getOperation()->getOperands().begin() + 2,
+                  getOperation()->getOperands().end());
+  Im2colOp tiledOp = mlir::clone(builder, *this,
+      TypeRange{outputSlice.getType()}, operands);
+  tiledOp.setMixedKOffset({kOffset});
+  tiledOp.setMixedMOffset({mOffset});
```

Comment:
IIRC, there is an assumption that the size of m_offset and k_offset are always one. I think we should return failure when it is not the case.

Reply:
This is now part of the op verifier.